REALMC-6371: add missing otto functionality

tsedgwick opened the pull request:
what is left to be added to this branch to make it usable for baas:
* memory
* properly handle stack trace 
see tests failing in baas

arahmanan commented on value.go:
```diff
@@ -53,7 +56,19 @@ var (
+}
+
 type Value interface {
+	ToInt() int
+	ToUInt32() uint32
+	ToInt32() int32
+	ToInt64() int64
```
[q] what's the reason for adding these functions to the interface?

arahmanan reviewed (commented):
Wow this is massive! I didn't go through it all yet, but I wanted to submit some of the questions I had so far.

arahmanan commented on added_values.go:
```diff
@@ -0,0 +1,608 @@
+}
+
+func (i valueUInt32) string() unistring.String {
+	return unistring.String(strconv.FormatInt(int64(i.ToInt()), 10))
```
[nit] `return unistring.String(i.String())`? This applies to the other `string()` functions below as well.

arahmanan commented on added_values.go:
```diff
@@ -0,0 +1,608 @@
+}
+
+func (i valueInt32) Equals(other Value) bool {
+	if o, ok := other.assertInt32(); ok {
```
This function looks identical to the `(i valueUInt32) Equals` function. Can me refactor this into a function called by both?

arahmanan commented on added_values.go:
```diff
@@ -0,0 +1,608 @@
+}
+
+func (i valueUInt32) Equals(other Value) bool {
+	if o, ok := other.assertInt32(); ok {
```
Do we need to check if `other` is a uint32 as well?

tsedgwick replied:
good catch - uint32 should be checking if the other equals uint32.

arahmanan commented on added_values.go:
```diff
@@ -0,0 +1,608 @@
+}
+
+func (i valueInt64) assertInt() (int, bool) {
+	return 0, true
```
Should we be returning `false` here?

tsedgwick commented on added_values.go:
```diff
@@ -0,0 +1,608 @@
+
+func (i valueNumber) ToInteger() int64 {
+	switch v := i.val.(type) {
+	case int:
```
yes - since we keep the int type in otto, I kept the values in here as well. This is useful for bson types Long vs Int for example.

tsedgwick replied:
When I was porting over goja from otto in baas, I didn't need those types so left them out. I can add them here, but ultimately it might be worth revisiting all of these added int types to the interface after we remove otto from baas.

makesitgo commented on vm.go:
```diff
@@ -2354,7 +2456,7 @@ func (_typeof) exec(vm *vm) {
 	case valueString:
 		r = stringString
-	case valueInt, valueFloat:
+	case valueInt, valueFloat, valueNumber:
```
would this case maybe also want the newest `value*` types?

makesitgo commented on vm.go:
```diff
@@ -2519,11 +2621,29 @@ type iterNext int32
 }
+
+func (stack valueStack) MemUsage(ctx *MemUsageContext) (uint64, error) {
+	total := uint64(0)
```
[personal style] more of a `var total uint64` guy myself

makesitgo commented on regexp_test.go:
```diff
@@ -489,9 +489,9 @@ func TestRegexpOverrideSpecies(t *testing.T) {
-func TestRegexp2InvalidEscape(t *testing.T) {
-	testScript1(`/(?=)\x0/.test("x0")`, valueTrue, t)
-}
+// func TestRegexp2InvalidEscape(t *testing.T) {
```
Intentional still?  Might be worth using a `t.Skip` instead?

makesitgo commented on runtime_test.go:
```diff
@@ -1357,61 +1356,63 @@ func TestCreateObject(t *testing.T) {
-	}
-}
+// (realm): we currently rely on this panicking back up to stop the execution.
+// in the future, we should look to be able to handle this correctly and remove out crutch on a panic.
```
[nit] is `out` a typo of `our`?

tsedgwick commented on array.go:
```diff
@@ -632,3 +632,26 @@ func strToGoIdx(s unistring.String) int {
+func (a *arrayObject) MemUsage(ctx *MemUsageContext) (uint64, error) {
+	total := EmptySize
+
+	// inc, err := a.baseObject.MemUsage(ctx)
```
maybe? memusage will be handled in a follow up ticket. since this won't be operational until we get that in, I didn't clean up any of this code: https://jira.mongodb.org/browse/REALMC-7452

tsedgwick commented on builtin_array.go:
```diff
@@ -1321,6 +1334,7 @@ func (a *arraySortCtx) sortCompare(x, y Value) int {
 
 	if a.compare != nil {
 		f := a.compare(FunctionCall{
+			ctx:       context.Background(),
```
good catch, meant to follow up on this one and add ctx to the arraySortCtx

makesitgo commented on vm.go:
```diff
@@ -1712,6 +1789,10 @@ func (_pop) exec(vm *vm) {
 }
 
 func (vm *vm) callEval(n int, strict bool) {
+	// (DIVERSION: REALMC-5102 if the user tries to call eval, we will throw a function not found
```
I get it now, thanks for the clarification

makesitgo commented on runtime.go:
```diff
@@ -1944,6 +2031,44 @@ func (r *Runtime) New(construct Value, args ...Value) (o *Object, err error) {
+type CallableWithContext func(ctx context.Context, this Value, args ...Value) (Value, error)
+
+func ToFunctionWithContext(v Value) (CallableWithContext, bool) {
+	if obj, ok := v.(*Object); ok {
```
makes sense to me!

makesitgo commented on string_ascii.go:
```diff
@@ -326,3 +326,110 @@ func (s asciiString) Export() interface{} {
+			return int64(f)
+		}
+	}
+	return i
```
:noice:

tsedgwick commented on builtin_object.go:
```diff
@@ -355,6 +355,58 @@ func (r *Runtime) object_keys(call FunctionCall) Value {
+	keys := object.Keys()
+	result := make([]Value, 0, len(keys))
+	for _, name := range keys {
+		result = append(result, r.newArrayValues([]Value{r.ToValue(name), object.Get(name)}))
```
this was a change that was added when I rebased on the current goja master. the main difference is that `Get` could end up calling `getStr` while `getOwnProp` will end up calling `getOwnProp`. `getStr` does the following for context: 
```
	prop := o.values[name]
	if prop == nil {
		if o.prototype != nil {
			if receiver == nil {
				return o.prototype.self.getStr(name, o.val)
			}
			return o.prototype.self.getStr(name, receiver)
		}
	}
	if prop, ok := prop.(*valueProperty); ok {
		if receiver == nil {
			return prop.get(o.val)
		}
		return prop.get(receiver)
	}
	return prop
```

`getOwnProp` will strictly return `o.values[name]`

makesitgo replied:
roger that, thanks for the explanation here!

makesitgo commented on vm.go:
```diff
@@ -301,10 +304,36 @@ func (vm *vm) run() {
+		} else {
+			vm.prg.code[vm.pc].exec(vm)
 		}
-		vm.prg.code[vm.pc].exec(vm)
```
🤦

tsedgwick commented on builtin_object.go:
```diff
@@ -355,6 +355,58 @@ func (r *Runtime) object_keys(call FunctionCall) Value {
+			panic(r.NewTypeError("Object must be of type array"))
+		}
+		key := entry.Get("0")
+		value := entry.Get("1")
```
it won't panic, it will return a nil

arahmanan commented on builtin_typedarrays.go:
```diff
@@ -1246,7 +1258,9 @@ func (r *Runtime) createArrayBufferProto(val *Object) objectImpl {
 		configurable: true,
+		writable:     true,
 		getterFunc:   r.newNativeFunc(r.arrayBufferProto_getByteLength, nil, "get byteLength", nil, 0),
+		setterFunc:   r.newNativeFunc(r.arrayBufferProto_getByteLength, nil, "set byteLength", nil, 0),
```
`arrayBufferProto_setByteLength` instead of `arrayBufferProto_getByteLength`?

arahmanan commented on builtin_typedarrays.go:
```diff
@@ -1276,7 +1299,9 @@ func (r *Runtime) createDataViewProto(val *Object) objectImpl {
 		configurable: true,
+		writable:     true,
 		getterFunc:   r.newNativeFunc(r.dataViewProto_getByteLen, nil, "get byteLength", nil, 0),
+		setterFunc:   r.newNativeFunc(r.dataViewProto_getByteLen, nil, "get byteLength", nil, 0),
```
We'll also want to pass in a setter function instead of `dataViewProto_getByteLen`. Same goes for `typedArrayProto_getByteLen` below.

tsedgwick commented on added_values.go:
```diff
@@ -0,0 +1,608 @@
+	return 0
+}
+
+func (i valueNumber) ToInt() int {
```
yes - since we keep the int type in otto, I kept the values in here as well. This is useful for bson types Long vs Int for example.

tsedgwick commented on mem_context.go:
```diff
@@ -0,0 +1,123 @@
+
+func (vt visitTracker) IsStashVisited(stash *stash) bool {
+	_, ok := vt.stashesVisited[stash]
+	// fmt.Println("visited :check:")
```
goes to overall conversation of mem usage, I can remove all this functionality. I left it in since this lib is unusable until we implement mem usage.

tsedgwick commented:
Added this project to evergreen and an .evg.yml - so we are good to go from that front. I believe I addressed all the comments. happy to revisit any or take it offline!

arahmanan reviewed (approved):
LGTM! Great work!

makesitgo commented on added_values.go:
```diff
@@ -0,0 +1,608 @@
+	if o, ok := other.assertInt(); ok {
+		return i.ToInt() == o
+	}
+	if o, ok := other.assertFloat(); ok {
```
I'm all for existing patterns... carry on here.  (just to clarify though, the suggestion here was just to leverage `SameAs` for the portion/subset of similar types being tested were the same, I wasn't suggesting they both keep the exact same implementations)

tsedgwick replied:
Got it! I misunderstood what you meant then. I thought you wanted me to replace the whole function with the SameAs implementation (`return other.SameAs(o)`).

makesitgo commented on added_values.go:
```diff
@@ -0,0 +1,608 @@
+	switch v := i.val.(type) {
+	case int:
+		return int64(v)
+	case int8:
```
Saw your answer to Andy's comment re: this, good enough for me!

makesitgo commented on mem_context.go:
```diff
@@ -0,0 +1,123 @@
+	return ok
+}
+
+// func (vt visitTracker) IsStackVisited(stash valueStack) bool {
```
Im guessing here this is the same situation as all the other commented out code related to mem usage, going to resolve this for now until we implement that portion

makesitgo reviewed (approved):
I went through my batch of comments and your changes... there's only one question and two nits left.  The question's probably at least worth answering, but otherwise this looks good to go from me

🆒  beans

tsedgwick commented on added_values.go:
```diff
@@ -0,0 +1,608 @@
+	if otherInt, ok := other.assertUInt32(); ok {
+		return uint32(i) == otherInt
+	}
+	return false
```
valueNumber was intended to be a bit more loose (any number) than these specific int types. So, I don't believe we should have the float comparisons here.
